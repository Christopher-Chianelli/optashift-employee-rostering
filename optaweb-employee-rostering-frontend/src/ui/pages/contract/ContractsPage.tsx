/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import {
  TableRow, TableCell, RowViewButtons, RowEditButtons,
  DataTableUrlProps, setSorterInUrl, TheTable,
} from 'ui/components/DataTable';
import { TextInput, Text } from '@patternfly/react-core';
import { useDispatch, useSelector } from 'react-redux';
import { Sorter } from 'types';
import { stringSorter } from 'util/CommonSorters';
import { useTranslation } from 'react-i18next';
import { RouteComponentProps, withRouter } from 'react-router';
import { tenantSelectors } from 'store/tenant';
import { useValidators } from 'util/ValidationUtils';
import { getPropsFromUrl } from 'util/BookmarkableUtils';
import { usePagableData } from 'util/FunctionalComponentUtils';
import { contractOperations, contractSelectors } from 'store/contract';
import { Contract } from 'domain/Contract';

export type Props = RouteComponentProps;

export const ContractRow = (contract: Contract) => {
  const [isEditing, setIsEditing] = useState(false);
  const dispatch = useDispatch();

  if (isEditing) {
    return (<EditableContractRow contract={contract} isNew={false} onClose={() => setIsEditing(false)} />);
  }

  return (
    <TableRow>
      <TableCell>
        <Text>{contract.name}</Text>
      </TableCell>
      <TableCell>
        <Text>{contract.maximumMinutesPerDay}</Text>
      </TableCell>
      <TableCell>
        <Text>{contract.maximumMinutesPerWeek}</Text>
      </TableCell>
      <TableCell>
        <Text>{contract.maximumMinutesPerMonth}</Text>
      </TableCell>
      <TableCell>
        <Text>{contract.maximumMinutesPerYear}</Text>
      </TableCell>
      <RowViewButtons
        onEdit={() => setIsEditing(true)}
        onDelete={() => dispatch(contractOperations.removeContract(contract))}
      />
    </TableRow>
  );
};

export const EditableContractRow = (props: { contract: Contract; isNew: boolean; onClose: () => void }) => {
  const [name, setName] = useState(props.contract.name);
  const [maximumMinutesPerDay, setMaximumMinutesPerDay] = useState(props.contract.maximumMinutesPerDay);
  const [maximumMinutesPerWeek, setMaximumMinutesPerWeek] = useState(props.contract.maximumMinutesPerWeek);
  const [maximumMinutesPerMonth, setMaximumMinutesPerMonth] = useState(props.contract.maximumMinutesPerMonth);
  const [maximumMinutesPerYear, setMaximumMinutesPerYear] = useState(props.contract.maximumMinutesPerYear);

  const dispatch = useDispatch();
  const contractList = useSelector(contractSelectors.getContractList);
  // const { t } = useTranslation("ContractsPage");

  const validators = {
    nameMustNotBeEmpty: {
      predicate: (contract: Contract) => contract.name.length > 0,
      errorMsg: () => 'Contract cannot have an empty name',
    },
    nameAlreadyTaken: {
      predicate: (contract: Contract) => contractList.filter(otherContract => otherContract.name === contract.name
        && otherContract.id !== contract.id).length === 0,
      errorMsg: (contract: Contract) => `Name (${contract.name}) is already taken by another contract`,
    },
  };

  const updatedContract: Contract = {
    ...props.contract,
    name,
    maximumMinutesPerDay,
    maximumMinutesPerWeek,
    maximumMinutesPerMonth,
    maximumMinutesPerYear,
  };

  const validationErrors = useValidators(updatedContract, validators);

  return (
    <TableRow>
      <TableCell>
        <TextInput value={name} onChange={setName} />
        {validationErrors.showValidationErrors('nameMustNotBeEmpty', 'nameAlreadyTaken')}
      </TableCell>
      <TableCell>
        <TextInput
          value={maximumMinutesPerDay || ''}
          onChange={(value) => {
            setMaximumMinutesPerDay(value ? parseInt(value, 10) : null);
          }}
          type="number"
          min={0}
        />
      </TableCell>
      <TableCell>
        <TextInput
          value={maximumMinutesPerWeek || ''}
          onChange={(value) => {
            setMaximumMinutesPerWeek(value ? parseInt(value, 10) : null);
          }}
          type="number"
          min={0}
        />
      </TableCell>
      <TableCell>
        <TextInput
          value={maximumMinutesPerMonth || ''}
          onChange={(value) => {
            setMaximumMinutesPerMonth(value ? parseInt(value, 10) : null);
          }}
          type="number"
          min={0}
        />
      </TableCell>
      <TableCell>
        <TextInput
          value={maximumMinutesPerYear || ''}
          onChange={(value) => {
            setMaximumMinutesPerYear(value ? parseInt(value, 10) : null);
          }}
          type="number"
          min={0}
        />
      </TableCell>
      <RowEditButtons
        isValid={validationErrors.isValid}
        onSave={() => {
          if (props.isNew) {
            dispatch(contractOperations.addContract(updatedContract));
          } else {
            dispatch(contractOperations.updateContract(updatedContract));
          }
        }}
        onClose={() => props.onClose()}
      />
    </TableRow>
  );
};

export const ContractsPage: React.FC<Props> = (props) => {
  const contractList = useSelector(contractSelectors.getContractList);
  const tenantId = useSelector(tenantSelectors.getTenantId);

  const { t } = useTranslation('ContractsPage');

  const columns = [
    { name: t('name'), sorter: stringSorter<Contract>(spot => spot.name) },
    { name: t('maxMinutesPerDay') },
    { name: t('maxMinutesPerWeek') },
    { name: t('maxMinutesPerMonth') },
    { name: t('maxMinutesPerYear') },
  ];

  const urlProps = getPropsFromUrl<DataTableUrlProps>(props, {
    page: '1',
    itemsPerPage: '10',
    filter: null,
    sortBy: '0',
    asc: 'true',
  });

  const sortBy = parseInt(urlProps.sortBy || '-1', 10);
  const sorter = columns[sortBy].sorter as Sorter<Contract>;

  const pagableData = usePagableData(urlProps, contractList, contract => [contract.name,
    `${contract.maximumMinutesPerDay || ''}`, `${contract.maximumMinutesPerWeek || ''}`,
    `${contract.maximumMinutesPerMonth || ''}`, `${contract.maximumMinutesPerYear || ''}`],
  sorter);

  return (
    <TheTable
      {...props}
      {...pagableData}
      title={t('contracts')}
      columns={columns}
      rowWrapper={contract => (<ContractRow key={contract.id} {...contract} />)}
      sortByIndex={sortBy}
      onSorterChange={index => setSorterInUrl(props, urlProps, sortBy, index)}
      newRowWrapper={removeRow => (
        <EditableContractRow
          isNew
          onClose={removeRow}
          contract={{
            tenantId,
            name: '',
            maximumMinutesPerDay: null,
            maximumMinutesPerWeek: null,
            maximumMinutesPerMonth: null,
            maximumMinutesPerYear: null,
          }}
        />
      )}
    />
  );
};

export default withRouter(ContractsPage);
