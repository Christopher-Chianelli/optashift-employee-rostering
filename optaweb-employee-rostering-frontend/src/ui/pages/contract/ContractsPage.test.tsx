/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Contract } from 'domain/Contract';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { mockStore } from 'store/mockStore';
import { Map } from 'immutable';
import { mockRedux } from 'setupTests';
import { ContractRow, ContractsPage, EditableContractRow } from './ContractsPage';

const noContractsStore = mockStore({
  contractList: {
    isLoading: false,
    contractMapById: Map(),
  },
}).store;

const twoContractsStore = mockStore({
  contractList: {
    isLoading: false,
    contractMapById: Map<number, Contract>()
      .set(0, {
        id: 0,
        version: 0,
        tenantId: 0,
        name: 'Contract 1',
        maximumMinutesPerDay: 1,
        maximumMinutesPerWeek: 2,
        maximumMinutesPerMonth: null,
        maximumMinutesPerYear: null,
      })
      .set(1,
        {
          id: 1,
          version: 0,
          tenantId: 0,
          name: 'Contract 2',
          maximumMinutesPerDay: null,
          maximumMinutesPerWeek: null,
          maximumMinutesPerMonth: 3,
          maximumMinutesPerYear: 4,
        }),
  },
}).store;

describe('Contracts page', () => {
  it('should render correctly with no skills', () => {
    mockRedux(noContractsStore);
    const skillsPage = shallow(<ContractsPage {...getRouterProps('/0/skill', {})} />);
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render correctly with a few skills', () => {
    mockRedux(twoContractsStore);
    const skillsPage = shallow(
      <ContractsPage {...getRouterProps('/0/skill', {})} />,
    );
    expect(toJson(skillsPage)).toMatchSnapshot();
  });

  it('should render the viewer correctly', () => {
    const contract = twoContractsStore.getState().contractList.contractMapById.get(0) as Contract;
    mockRedux(twoContractsStore);
    const viewer = shallow(<ContractRow {...contract} />);
    expect(toJson(viewer)).toMatchSnapshot();
  });

  it('should render the editor correctly', () => {
    const contract = twoContractsStore.getState().contractList.contractMapById.get(0) as Contract;
    mockRedux(twoContractsStore);
    const editor = shallow(<EditableContractRow contract={contract} isNew={false} onClose={jest.fn()} />);
    expect(toJson(editor)).toMatchSnapshot();
  });
});
