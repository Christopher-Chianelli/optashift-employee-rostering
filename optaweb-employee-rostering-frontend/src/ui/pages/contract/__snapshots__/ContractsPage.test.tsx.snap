// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contracts page should render correctly with a few skills 1`] = `
<TheTable
  columns={
    Array [
      Object {
        "name": "Trans(i18nKey=name)",
        "sorter": [Function],
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerDay)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerWeek)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerMonth)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerYear)",
      },
    ]
  }
  filterText=""
  filteredRows={
    Immutable.List [
      Object {
        "id": 0,
        "maximumMinutesPerDay": 1,
        "maximumMinutesPerMonth": null,
        "maximumMinutesPerWeek": 2,
        "maximumMinutesPerYear": null,
        "name": "Contract 1",
        "tenantId": 0,
        "version": 0,
      },
      Object {
        "id": 1,
        "maximumMinutesPerDay": null,
        "maximumMinutesPerMonth": 3,
        "maximumMinutesPerWeek": null,
        "maximumMinutesPerYear": 4,
        "name": "Contract 2",
        "tenantId": 0,
        "version": 0,
      },
    ]
  }
  history={
    Object {
      "action": "PUSH",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 1,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "/0/skill",
        "search": "",
        "state": undefined,
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
  isReversed={true}
  itemsPerPage={10}
  location={
    Object {
      "hash": "",
      "pathname": "/0/skill",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/0/skill",
      "url": "localhost:8080/0/skill",
    }
  }
  newRowWrapper={[Function]}
  numOfFilteredRows={2}
  onSorterChange={[Function]}
  page={1}
  rowWrapper={[Function]}
  rowsInPage={
    Immutable.List [
      Object {
        "id": 0,
        "maximumMinutesPerDay": 1,
        "maximumMinutesPerMonth": null,
        "maximumMinutesPerWeek": 2,
        "maximumMinutesPerYear": null,
        "name": "Contract 1",
        "tenantId": 0,
        "version": 0,
      },
      Object {
        "id": 1,
        "maximumMinutesPerDay": null,
        "maximumMinutesPerMonth": 3,
        "maximumMinutesPerWeek": null,
        "maximumMinutesPerYear": 4,
        "name": "Contract 2",
        "tenantId": 0,
        "version": 0,
      },
    ]
  }
  sortByIndex={0}
  title="Trans(i18nKey=contracts)"
/>
`;

exports[`Contracts page should render correctly with no skills 1`] = `
<TheTable
  columns={
    Array [
      Object {
        "name": "Trans(i18nKey=name)",
        "sorter": [Function],
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerDay)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerWeek)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerMonth)",
      },
      Object {
        "name": "Trans(i18nKey=maxMinutesPerYear)",
      },
    ]
  }
  filterText=""
  filteredRows={Immutable.List []}
  history={
    Object {
      "action": "PUSH",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 1,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "/0/skill",
        "search": "",
        "state": undefined,
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
  isReversed={true}
  itemsPerPage={10}
  location={
    Object {
      "hash": "",
      "pathname": "/0/skill",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/0/skill",
      "url": "localhost:8080/0/skill",
    }
  }
  newRowWrapper={[Function]}
  numOfFilteredRows={0}
  onSorterChange={[Function]}
  page={1}
  rowWrapper={[Function]}
  rowsInPage={Immutable.List []}
  sortByIndex={0}
  title="Trans(i18nKey=contracts)"
/>
`;

exports[`Contracts page should render the editor correctly 1`] = `
<TableRow>
  <TableCell>
    <TextInput
      onChange={[Function]}
      value="Contract 1"
    />
  </TableCell>
  <TableCell>
    <TextInput
      min={0}
      onChange={[Function]}
      type="number"
      value={1}
    />
  </TableCell>
  <TableCell>
    <TextInput
      min={0}
      onChange={[Function]}
      type="number"
      value={2}
    />
  </TableCell>
  <TableCell>
    <TextInput
      min={0}
      onChange={[Function]}
      type="number"
      value=""
    />
  </TableCell>
  <TableCell>
    <TextInput
      min={0}
      onChange={[Function]}
      type="number"
      value=""
    />
  </TableCell>
  <RowEditButtons
    isValid={true}
    onClose={[Function]}
    onSave={[Function]}
  />
</TableRow>
`;

exports[`Contracts page should render the viewer correctly 1`] = `
<TableRow>
  <TableCell>
    <Text>
      Contract 1
    </Text>
  </TableCell>
  <TableCell>
    <Text>
      1
    </Text>
  </TableCell>
  <TableCell>
    <Text>
      2
    </Text>
  </TableCell>
  <TableCell>
    <Text />
  </TableCell>
  <TableCell>
    <Text />
  </TableCell>
  <RowViewButtons
    onDelete={[Function]}
    onEdit={[Function]}
  />
</TableRow>
`;
